src/app/clusters/ota-requestor/BDXDownloader.cpp

```diff
@@ -18,101 +19,200 @@
+            mImageProcessor->Abort();
         }
-        break;
-    case TransferSession::OutputEventType::kMsgToSend: {
-        chip::Messaging::SendFlags sendFlags;
-        VerifyOrReturn(mExchangeCtx != nullptr, ChipLogError(BDX, "mExchangeContext is null, cannot proceed"));
-        if (event.msgTypeData.MessageType == static_cast<uint8_t>(MessageType::ReceiveInit))
+        mState = State::kIdle;
+    }
+    else
+    {
+        ChipLogError(BDX, "no download in progress");
+    }
+}
+
+void BDXDownloader::EndDownload(CHIP_ERROR reason)
+{
+    if (mState == State::kInProgress)
+    {
+        // There's no method for a BDX receiving driver to cleanly end a transfer
+        mBdxTransfer.AbortTransfer(chip::bdx::StatusCode::kUnknown);
+        if (mImageProcessor != nullptr)
         {
-            sendFlags.Set(chip::Messaging::SendMessageFlags::kFromInitiator);
+            mImageProcessor->Abort();
         }
-        if (event.msgTypeData.MessageType != static_cast<uint8_t>(MessageType::BlockAckEOF))
+        mState = State::kIdle;
+
+        // Because AbortTransfer() will generate a StatusReport to send.
+        PollTransferSession();
+    }
+    else
+    {
+        ChipLogError(BDX, "no download in progress");
+    }
+}
+
+void BDXDownloader::PollTransferSession()
+{
+    TransferSession::OutputEvent outEvent;
+
+    // WARNING: Is this dangerous? What happens if the loop encounters two messages that need to be sent? Does the ExchangeContext
+    // allow that?
+    do
+    {
+        mBdxTransfer.PollOutput(outEvent, /* TODO:(#12520) */ chip::System::Clock::Seconds16(0));
+        CHIP_ERROR err = HandleBdxEvent(outEvent);
+        VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogError(BDX, "HandleBDXEvent: %" CHIP_ERROR_FORMAT, err.Format()));
+    } while (outEvent.EventType != TransferSession::OutputEventType::kNone);
+}
+
+CHIP_ERROR BDXDownloader::HandleBdxEvent(const chip::bdx::TransferSession::OutputEvent & outEvent)
+{
+    VerifyOrReturnError(mImageProcessor != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    switch (outEvent.EventType)
+    {
+    case TransferSession::OutputEventType::kNone:
+        break;
+    case TransferSession::OutputEventType::kAcceptReceived:
+        ReturnErrorOnFailure(mBdxTransfer.PrepareBlockQuery());
+        // TODO: need to check ReceiveAccept parameters
+        break;
+    case TransferSession::OutputEventType::kMsgToSend: {
+        VerifyOrReturnError(mMsgDelegate != nullptr, CHIP_ERROR_INCORRECT_STATE);
+        ReturnErrorOnFailure(mMsgDelegate->SendMessage(outEvent));
+        if (outEvent.msgTypeData.HasMessageType(chip::bdx::MessageType::BlockAckEOF))
         {
-            sendFlags.Set(chip::Messaging::SendMessageFlags::kExpectResponse);
+            // BDX transfer is not complete until BlockAckEOF has been sent
+            mState = State::kComplete;
+
+            // TODO: how/when to reset the BDXDownloader to be ready to handle another download
```

Comment:
What is everything involved to reset to handle another download? What flags/member vars need to be reset?

Reply:
That's not quite what I meant here. You can see below that there's logic for resetting:
```
        mBdxTransfer.Reset();
        ReturnErrorOnFailure(mImageProcessor->Abort());
```
What I meant was - specifically after the download completes successfully, should we automatically reset everything? Or do we remain in a `Completed` state and wait for the application or something else to reset it? My gut says we do the latter and let some external entity trigger the reset.

Reply:
I see. Perhaps before the start of a transfer, `Reset` should always be called to make sure we're in a good state.